### ISSUE TITLE
Using smart pointers for Pointclouds and Octrees in OctoMap

### ISSUE THREAD

**Issue by BhargavaRamM** [OctoMap/octomap#177]
@ahornung Thanks for creating a great library. Currently, I am using OctoMap in a project related to object detection in point clouds. 
I am wondering if there is any reason that I am not familiar with why smart pointers are not used in OctoMap? I used PCL before and liked the feature of smart pointer for a PointCloud.  I would like this option as an enhancement to OctoMap.

**Comment by BhargavaRamM** [comment]
@ahornung I tried implementing smart pointers for `octomap::Pointcloud` and `octomap::OcTree` in a crude way just to test if it works an it works fine.

If you think this is a good feature/enhancement, let me know and I am more than happy to help with coding.

**Comment by de-vri-es** [comment]
In the post C++11 era, smart pointers are included in the standard library. Using shared pointers is as easy as `std::shared_ptr<octomap::OcTree>`, or `auto octree = std::make_shared<octomap::Octree>("/path/to/file")`.

Even when a function returns an `OcTree` by value, move semantics allow wrapping it in a smart pointer without expensive data copying:
```
octomap::OcTree foo();
auto octree = std::make_shared<octomap::OcTree>(foo()); // No copy overhead with move support.
```

Functions should not take objects by shared pointer, unless they need to take (shared) ownership of the object. That is actually a pretty rare case, so normally the right thing to do is take a const/non-const ref, or even a copy.

If a function takes an object by shared pointer for no good reason, the library needlessly forces all objects to live in shared pointers. At best that can be a nuisance for the user of the library, in the worst case it can lead to performance problems or even incompatibles. For example, mixing `boost::` and `std::shared_ptrs` is not trivial, and cannot be done perfectly when `weak_ptrs` enter in the mix.

In other words, the less a library concerns itself with the way object lifetime is managed, the better, because it allows the user of the library the freedom to choose the lifetime management that best suits the application.

**Comment by BhargavaRamM** [comment]
@de-vri-es Sorry for very late reply. I was looking for a reason for why OctoMap is not using smart pointers. Thank you for the explanation. 
Like you mentioned, using `shared_ptr` is simple.
